File: Concurrency/1117_Building_H2O/Go/Solution.go
package main

import (
	"fmt"
	"sync"
)

// function type for passing function to methods
type ReleaseFunc func()

type H2O struct {
	// hydrogen molecule count
	H2Count int
	// mutex for r/w synchronization of hydrogen molecule count
	Mtx *sync.Mutex
}

// allocate & initiate channels
func (h2o *H2O) Init() {
	// lock mutex
	h2o.Mtx.Lock()
	// initialize hydrogen count to zero
	h2o.H2Count = 0
	// unlock mutex
	h2o.Mtx.Unlock()
}

// hydrogen molecule function
func (h2o *H2O) Hydrogen(releaseHydrogen ReleaseFunc) {
	// loop exit condition
	exitLoop := false

	for !exitLoop {
		// lock mutex
		h2o.Mtx.Lock()
		if h2o.H2Count < 2 {
			// increment hydrogen count if space available
			h2o.H2Count++
			// release hydrogen molecule
			releaseHydrogen()
			// exit loop
			exitLoop = true
		}
		// unlock mutex
		h2o.Mtx.Unlock()
	}
}

// oxygen molecule function
func (h2o *H2O) Oxygen(releaseOxygen ReleaseFunc) {
	// loop exit condition
	exitLoop := false

	for !exitLoop {
		// lock mutex
		h2o.Mtx.Lock()
		if h2o.H2Count >= 2 {
			// reset hydrogen count if water can be formed
			h2o.H2Count = 0
			// release oxygen molecule
			releaseOxygen()
			// exit loop
			exitLoop = true
		}
		// unlock mutex
		h2o.Mtx.Unlock()
	}
}

func main() {
	// input test string
	input := "OOOOHHHHHHHH"
	// mutex for r/w synchronization of hydrogen molecule count
	mtx := sync.Mutex{}
	// create wait group for goroutine synchronization
	wg := sync.WaitGroup{}
	// add one wait group for each letter
	wg.Add(len(input))

	// create & initiate H2O object
	h2o := H2O{
		Mtx: &mtx,
	}
	h2o.Init()

	// printer function for releasing hydrogen
	releaseH := func() {
		fmt.Print("H")
	}
	// printer function for releasing oxygen
	releaseO := func() {
		fmt.Print("O")
	}

	var char string
	for _, c := range input {
		// convert rune to character string
		char = fmt.Sprintf("%c", c)
		if char == "H" {
			// hydrogen goroutine
			go func() {
				h2o.Hydrogen(releaseH)
				// signal wait group termination
				wg.Done()
			}()
		} else if char == "O" {
			// oxygen goroutine
			go func() {
				h2o.Oxygen(releaseO)
				// signal wait group termination
				wg.Done()
			}()
		} else {
			panic("Invalid character found, `input` must consist of 'H' or 'O' characters")
		}
	}

	// wait for all waitgroups to be done
	wg.Wait()
	fmt.Print("\n")
}

File: Concurrency/1117_Building_H2O/Python/Solution.py
import threading

class H2O:
    def __init__(self):
        # hydrogen molecule count
        self.hydrogen_count = 0
        # oxygen molecule count
        self.oxygen_count = 0
        # mutex for r/w synchronization of hydrogen/oxygen counts
        self.mtx = threading.Lock()

    # bond hydrogen & oxygen molecules if possible
    def bond(self):
        # lock mutex
        self.mtx.acquire()
        # if water can be formed, clear hydrogen & oxygen counts
        if self.hydrogen_count == 2 and self.oxygen_count == 1:
            self.hydrogen_count = 0
            self.oxygen_count = 0
        # unlock mutex
        self.mtx.release()

    # hydrogen molecule function
    def hydrogen(self, releaseHydrogen: 'Callable[[], None]') -> None:
        # loop exit condition
        exit_loop = False

        while not exit_loop:
            # lock mutex
            self.mtx.acquire()
            if self.hydrogen_count < 2:
                # increment hydrogen count if space available
                self.hydrogen_count += 1
                # releaseHydrogen() outputs "H". Do not change or remove this line.
                releaseHydrogen()
                # exit loop
                exit_loop = True

            # unlock mutex
            self.mtx.release()

        # bond molecules if possible
        self.bond()

    # oxygen molecule function
    def oxygen(self, releaseOxygen: 'Callable[[], None]') -> None:
        # loop exit condition
        exit_loop = False

        while not exit_loop:
            # lock mutex
            self.mtx.acquire()
            if self.oxygen_count < 1:
                # increment oxygen count if space available
                self.oxygen_count += 1
                # releaseOxygen() outputs "O". Do not change or remove this line.
                releaseOxygen()
                # exit loop
                exit_loop = True

            # unlock mutex
            self.mtx.release()

        # bond molecules if possible
        self.bond()